core/src/main/java/io/dekorate/BuildImage.java

```diff
@@ -0,0 +1,149 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ **/
+
+package io.dekorate;
+
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.ToIntFunction;
+import java.util.regex.Pattern;
+
+import io.dekorate.utils.Strings;
+
+public class BuildImage {
+
+  private static final String DOT = "[\\._\\-]+";
+  private static final String EMPTY = "";
+  private static final String NON_DIGIT = "\\D+";
+
+  private final String image;
+  private final String[] command;
+
+  private final String tool;
+  private final String toolVersion;
+  private final int jdkVersion;
+  private final String jdkFlavor;
+
+  private static final List<BuildImage> IMAGES = Arrays.asList(
+                                                                   new BuildImage("docker.io/maven:3.6.3-jdk-8", "maven", "3.6.3", 8, "openjdk", "mvn", "clean", "install"),
+                                                                   new BuildImage("docker.io/maven:3.6.3-jdk-11", "maven", "3.6.3", 11, "openjdk", "mvn", "clean", "install" ),
+                                                                   new BuildImage("docker.io/maven:3.6.3-jdk-13", "maven", "3.6.3", 14, "openjdk", "mvn", "clean", "install"),
+
+                                                                   new BuildImage("docker.io/maven:3.6.3-amazoncorretto-8", "maven", "3.6.3", 8, "amazoncorretto", "mvn", "clean", "install"),
+                                                                   new BuildImage("docker.io/maven:3.6.3-amazoncorretto-11", "maven", "3.6.3", 11, "amazoncorreto", "mvn", "clean", "install"),
+
+                                                                   new BuildImage("docker.io/maven:3.6.3-openj9-11", "maven", "3.6.3", 11, "openj9", "mvn", "clean", "install"),
+                                                                   new BuildImage("docker.io/maven:3.6.3-ibmjava-8", "maven", "3.6.3", 8, "ibmjava", "mvn", "clean", "install"),
+
+                                                                   new BuildImage("docker.io/gradle:6.3.0-jdk8", "gradle", "6.3.0", 8, "openjdk", "gradle", "clean", "build"),
+                                                                   new BuildImage("docker.io/gradle:6.3.0-jdk11", "gradle", "6.3.0", 11, "openjdk", "gradle", "clean", "build"),
+                                                                   new BuildImage("docker.io/gradle:6.3.0-jdk13", "gradle", "6.3.0", 13, "openjdk", "gradle", "clean", "build")
+  );
+
+  public static Optional<BuildImage> find(String tool, String toolVersion, int jdkVersion, String jdkFlavor) {
```

Comment:
IIUC this get's the closest matching build image?

If so, it should be documented - I mean you are using geometry to find the closest match, so we need at least need to have the rationale documented :)